core/shared/src/main/scala/zio/stm/TArray.scala

```diff
@@ -26,6 +26,65 @@ class TArray[A] private (val array: Array[TRef[A]]) extends AnyVal {
     if (0 <= index && index < array.length) array(index).get
     else STM.die(new ArrayIndexOutOfBoundsException(index))
 
+  /** Finds the result of applying a partial function to the first value in its domain. */
+  final def collectFirst[B](pf: PartialFunction[A, B]): STM[Nothing, Option[B]] =
+    find(pf.isDefinedAt).map(_.map(pf))
+
+  /** Finds the result of applying an STM partial function to the first value in its domain. */
+  final def collectFirstM[E, B](pf: PartialFunction[A, STM[E, B]]): STM[E, Option[B]] =
+    find(pf.isDefinedAt).flatMap {
+      case Some(a) => pf(a).map(Some(_))
+      case _       => STM.succeed(None)
+    }
+
+  /** Determine if the array contains a specified value. */
+  final def contains(a: A): STM[Nothing, Boolean] = exists(_ == a)
+
+  /** Count the values in the array matching a predicate. */
+  final def count(predicate: A => Boolean): STM[Nothing, Int] = fold(0) { (n, a) =>
+    if (predicate(a)) n + 1 else n
+  }
+
+  /** Count the values in the array matching an STM predicate. */
+  final def countM[E](predicate: A => STM[E, Boolean]): STM[E, Int] = foldM[E, Int](0) { (n, a) =>
+    predicate(a).map(result => if (result) n + 1 else n)
+  }
+
+  /** Determine if the array contains a value satisfying a predicate. */
+  final def exists(predicate: A => Boolean): STM[Nothing, Boolean] = find(predicate).map(_.isDefined)
+
+  /** Determine if the array contains a value satisfying an STM predicate. */
+  final def existsM[E](predicate: A => STM[E, Boolean]): STM[E, Boolean] =
+    countM(predicate).map(_ > 0) //Existence should not be order dependent so the entire array needs to be observed.
+
+  /** Find the first element in the array matching a predicate. */
+  final def find(predicate: A => Boolean): STM[Nothing, Option[A]] =
+    if (array.isEmpty) STM.succeed(None)
```

Comment:
Maybe express it in terms of `fold`?

Reply:
The reason I did not was so that it could return early at the first match. TArray doesn't have a foldWhile (like Chunk, for example). I'm happy to change it to use fold if you think it's better to keep it simpler. What do you think?

Reply:
Good point. Let's keep it this way, and add the combinators like foldWhile once we make an iteration to optimizer the transactional structures.

Reply:
Good point. Let's keep it this way, and add the combinators like foldWhile once we make an iteration to optimizer the transactional structures.